### DISCUSSION TITLE
creating parameter file in docker installation

### DISCUSSION THREAD

**Discussion by krishnakumarkkp** [davidfrantz/force#122] [Q&A] [answered]
In docker installation, while using command <sudo docker run davidfrantz/force force-parameter /data/force/param LEVEL2 1>, I am getting the following error: 

too many non-optional arguments.
Usage: force-parameter [-h] [-v] [-i] [-m] [-c] parameter-file module

**Comment by Florian-Katerndahl** [comment]
Hey @krishnakumarkkp,
in FORCE 3.7.0, the verbosity of parameter files is not given with an suffixed integer but is controlled via the `-c` flag ([see latest documentation for `force-parameter`)](https://force-eo.readthedocs.io/en/latest/components/auxilliary/parameter.html). Setting `-c` creates a compact parameter file, while the default is a verbose one.

**Reply by krishnakumarkkp** [reply]
But they are optional parameters. I just want to check that what is wrong with the command <sudo docker run davidfrantz/force force-parameter /data/force/param LEVEL2 1> as it is throwing error 'too many non-optional arguments'. 

And when I omit the '1' at the end, I am getting an error message: "Unable to open output file!"

**Reply by ernstste** [reply]
Check the usage again:
`force-parameter [-h] [-v] [-i] [-m] [-c] parameter-file module`
and compare it with your use:
`force force-parameter /data/force/param LEVEL2 1`

Correct usage: The first argument is the parameter file (output), the second argument is the module (LEVEL2 in your case). 
The way you're trying to use it: First argument: module, second argument: verbosity (which is now controlled by the `-c` parameter, see Florian's response)

Hope that helps

**Reply by Florian-Katerndahl** [reply]
The command should be: `sudo docker run davidfrantz/force force-parameter /data/force/param/l2ps.prm LEVEL2` for a verbose version and `sudo docker run davidfrantz/force force-parameter -c /data/force/param/l2ps.prm LEVEL2` for the short parameter file.
The traling `1` should not be included and it is neccessary to specifiy the filename as well.
Edit: The comment this was supposed to answer doesn't show up anymore. Was it deleted?

**Reply by krishnakumarkkp** [reply]
Yes. I got it. I was supposed to ask the reason for the error message: "Unable to open output file!"' when executing the correct command

ps; Sorry for deleting the previous comment. Came out of misunderstanding

**Comment by davidfrantz** [comment, marked as answer]
Hi @krishnakumarkkp ,

@Florian-Katerndahl is spot-on.

To elaborate a bit: the usage changed in v. 3.7 to a more Unix-style usage using options. The trailing ``1`` is now replaced by the ``-c`` (compact) option. Also, you now have to specify the output file name, which eliminates the necessity to rename the generated parameter file afterwards (and prevents you from accidentally overwriting previous parameter files when not renaming the files).

I assume that you are referring to some of the tutorials. There is always an indication for the version number for which the tutorial was made. New versions might need some adaptation (as in this case).

Cheers,
David

**Reply by krishnakumarkkp** [reply]
Thank you for your answer. I got it right. But still getting an error message "Unable to open output file!"'

**Reply by davidfrantz** [reply]
Is ``/data/force/param`` a directory?

**Reply by krishnakumarkkp** [reply]
Yes. I tried the same command by replacing it with a different directory:  '/home/Krishna/mydata/param'

**Reply by Florian-Katerndahl** [reply]
As stated by @davidfrantz and myself above, you  need to specify the filename in v3.7 since FORCE doesn't generate placeholder-names anymore.
Edit: a valid path could be `/home/Krishna/mydata/param/my-l2ps-file.prm`.

**Reply by krishnakumarkkp** [reply]
I tried with both options (directory and/or file name). Still getting the same error: "Unable to open output file!"'

Like : `sudo docker run davidfrantz/force force-parameter /home/Krishna/mydata/param/my-l2ps-file.prm LEVEL2`

**Reply by Florian-Katerndahl** [reply]
Does the directory exist, though? When trying to write to a file, whose path ist not valid, e.g. the folder `mydata` is missing, FORCE throws the same error as if no filename was given. 
If that's not the problem, I don't have anyore ideas on what to do.

**Reply by krishnakumarkkp** [reply]
Yes. The directory exist and I am able to create a new empty text file there using the 'cat' command.

**Reply by krishnakumarkkp** [reply]
OK. I created a skelton parameter file using

`sudo docker run davidfrantz/force force-parameter my-l2ps-file.prm LEVEL2`

without providing the path (directory). I don't know the actual location of the file now

**Reply by davidfrantz** [reply]
Could be a Docker issue then.

What about simply creating an empty file with standard Unix-commands through the docker container?

``docker run davidfrantz/force touch /home/Krishna/mydata/param/my-l2ps-file.prm``

Not sure why you work with sudo, this might be part of the problem, too?

**Reply by davidfrantz** [reply]
Alright, the problem is your usage of Docker then... This is no FORCE-specific issue.

Please carefully read through https://force-eo.readthedocs.io/en/latest/setup/docker.html

You have to mount some folders and make sure that the user within the Docker container is allowed to your host file system.

**Reply by krishnakumarkkp** [reply]
Will check this and let you know. Thank you!

**Reply by krishnakumarkkp** [reply]
I was trying to map a local folder to a folder within the container using:

`docker run -v /home/Krishna:/opt/data davidfrantz/force force-level2 /opt/data/parameters.prm`

It then throws "/opt/data/parameters.prm does not exist". 

is this '/opt/data/' is the default directory where parameters.prm is stored?

[As per the tutorial in https://force-eo.readthedocs.io/en/latest/setup/docker.html]

Update:

Now I have a parameter file at this location: /home/Krishna/param.prm and have data at the same location /home/Krishna/*.tif

I generated this parameter file using:

`docker run -v /home/Krishna:/opt/data davidfrantz/force force-parameter /opt/data/param.prm LEVEL2`

which created an empty Skelton at /home/Krishna and then I edited this using the vi editor.

Now I just want to run the command force-l2ps to produce the level2 data. 
But running the command:

`docker run davidfrantz/force force-l2ps /home/Krishna /home/Krishna/param.prm`

throws the error

"Unable to open parameter file!
Parsing parameter file failed."

**Reply by davidfrantz** [reply]
Hey, 
Can you please upload the parameter file, and additionally paste the output of the following commands here?
``id``
``ls -l /home/Krishna/param.prm``

**Reply by krishnakumarkkp** [reply]
`id  `
uid=1000(ubuntu) gid=1000(ubuntu) groups=1000(ubuntu),4(adm),20(dialout),24(cdrom),25(floppy),27(sudo),29(audio),30(dip),44(video),46(plugdev),117(netdev),118(lxd),998(docker)
`ls -l /home/Krishna/param.prm`
-rw-r--r-- 1 ubuntu ubuntu 16383 Sep  8 07:08 /home/Krishna/param.prm

**Reply by krishnakumarkkp** [reply]
I am pretty sure that there is nothing wrong with the parameter file, as I have tested the same command with a skeleton parameter file when I got the same error "Unable to open parameter file! Parsing parameter file failed." 

But when I created and tested with a  skeleton parameter file at '/opt/data/LEVEL2-skeleton.prm', it throws error something like: 'parameter --- is NULL and parameter --- is NULL'

So I guess the reason for the initial error may be because it does not have proper access to the parameter file, though I have mapped the local directory right.

ps: 
`docker run -v /home/Krishna:/opt/data --user "$(id -u):$(id -g)" davidfrantz/force force-parameter /opt/data/testparam.prm LEVEL2`

gives me a similar error: "Unable to open output file!"

and at the same time
`docker run -v /home/Krishna:/opt/data --user "$(id -u):$(id -g)" davidfrantz/force force`
displays the correct output

**Reply by davidfrantz** [reply]
The program ``force`` does not write output, which is probably pointing to some Docker configuration issue that is not resolved yet.

I am a bit confused now as you wrote earlier that the generation of the parameter file worked before. Let's try if we can get to pinpointing the issue using regular Unix-commands from within the container.

Who is the user within the container?
``docker run -v /home/Krishna:/opt/data --user "$(id -u):$(id -g)" davidfrantz/force id``

Can we write files through the container?
``docker run -v /home/Krishna:/opt/data --user "$(id -u):$(id -g)" davidfrantz/force touch opt/data/test``

Can we see the mounted directory and files in the container (make sure to delete private lines before posting, this shows the content of your home directory)
``docker run -v /home/Krishna:/opt/data --user "$(id -u):$(id -g)" davidfrantz/force ls -l opt/data``

Can we see the created test file on the host?
``ls -l /home/Krishna/test``

**Reply by krishnakumarkkp** [reply]
It is true that the generation of the parameter file worked before. I have used the following command to create a parameter file.
`docker run -v /home/Krishna:/opt/data davidfrantz/force force-parameter /opt/data/param.prm LEVEL2`
Now the same command throwing error "Unable to open output file!"
But when I do:
`docker run -v /home/Krishna/mydata:/opt/data davidfrantz/force force-parameter /opt/data/param.prm LEVEL2`
it created a parameter file at /home/Krishna/mydata/param.prm
but 
`docker run davidfrantz/force force-l2ps /home/Krishna/mydata /home/Krishna/mydata/param.prm` 
again throwing error "Unable to open parameter file!
Parsing parameter file failed."

*user within the container?
`docker run -v /home/Krishna:/opt/data --user "$(id -u):$(id -g)" davidfrantz/force id`

produces
output: uid=1000(docker) gid=1000(docker) groups=1000(docker)
-------------------------------------------------------------------------------------------------------------------
*Can we write files through the container?
`docker run -v /home/Krishna:/opt/data --user "$(id -u):$(id -g)" davidfrantz/force touch opt/data/test`

output: touch: cannot touch 'opt/data/test': No such file or directory
--------------------------------------------------------------------------------------------------------------------
*Can we see the mounted directory and files in the container (make sure to delete private lines before posting, this shows the content of your home directory)
`docker run -v /home/Krishna:/opt/data --user "$(id -u):$(id -g)" davidfrantz/force ls -l opt/data`

output: ls: cannot access 'opt/data': No such file or directory
--------------------------------------------------------------------------------------------------------------------
Can we see the created test file on the host?
`ls -l /home/Krishna/test`

output: ls: cannot access '/home/Krishna/test': No such file or directory

**The only thing that I found working is :**

`docker run -v /home/Krishna/mydata:/opt/data davidfrantz/force force-parameter /opt/data/test_parameters2.prm LEVEL2`

**says: "An empty parameter file skeleton was written to  /opt/data/test_parameters2.prm"**

**Reply by davidfrantz** [reply]
Hi,
sorry, I had a spelling error in my previous response. It should be ``/opt/data`` in all cases, not ``opt/data``
Sorry...

However, there is already some mistake here as the container doesn't know the file structure on the host:
``docker run davidfrantz/force force-l2ps /home/Krishna/mydata /home/Krishna/mydata/param.prm``

Change this to:
``docker run  -v /home/Krishna/mydata:/opt/data davidfrantz/force force-l2ps /opt/data/ /opt/data/param.prm``

This should find your parameterfile, but will fail nonetheless, as you need to give a proper input image, too. E.g.:
``docker run  -v /home/Krishna/mydata:/opt/data davidfrantz/force force-l2ps /opt/data/LC08_L1GT_029030_20151209_20160131_01_T1 /opt/data/param.prm``

**Reply by krishnakumarkkp** [reply]
Yes. It worked. A lot of thanks for that! It finds the parameter file perfectly! 
I have my image data in the directory /home/Krishna/mydata.  But now when I execute
`docker run -v /home/Krishna/mydata:/opt/data davidfrantz/force force-l2ps /opt/data/BLR_L5_C2_T1_L2_2011 /opt/data/parameters.prm`

it gives me errors:
Couldn't open the directory: No such file or directory
Couldn't open the directory: No such file or directory
Couldn't open the directory: No such file or directory

I assume this means the output directories. However, I have mentioned them in the parameter file like:

DIR_LEVEL2 = /opt/data
DIR_LOG = /opt/data
DIR_TEMP = /opt/data

and FILE_QUEUE = NULL

**Reply by davidfrantz** [reply]
okay, we are almost there ;)

I would assume that this is rather related to the input. Is this directory in mydata in your home directory?
What image is this? FORCE can only work on Landsat and Sentinel-2 Level 1 images, and the input directories should be named in the usual way, e.g. ``/opt/data/LC08_L1TP_029030_20151209_20160131_01_T1``

Nevertheless, it might be better to use a separate folder for the output (DIR_LEVEL2), e.g. /opt/data/level2 (which needs to exist)

**Reply by krishnakumarkkp** [reply]
**/home/Krishna/mydata** is my default home directory in the file system.
 I already have following empty directories inside this parent directory: level2, temp, log.
And I have Landsat-5 level-1 images (16 bit) with 7 bands and the parameter file inside like:

 /home/Krishna/mydata/parameters.prm
/home/Krishna/mydata/BLR_L5_C2_T1_L2_2011.tif

initially I tried with separate directories, but got the same error above.
DIR_LEVEL2 = /opt/data/level2
DIR_LOG = /opt/data/log
DIR_TEMP = /opt/data/temp

Let me try with a different image

update: I tried with a different image 'LT05_L1TP_144051_20110203_20161010_01_T1_Stack_2.tif'. [this is a band stacked image with 7 bands (8 bit image)] This is a Landsat-5 level-1 image. But got the same set of errors.

**Reply by davidfrantz** [reply]
@krishnakumarkkp, 
this is because you have the wrong input data. Please download the data from USGS, untar the package, do not rename anything, do not stack anything
FORCE is meant to work with the original data

**Reply by davidfrantz** [reply]
or use force-level2, in which case you can even work with the tar archive without unpacking

**Reply by krishnakumarkkp** [reply]
Yes you are right.
Now with original data from USGS got the following output:

dc:  70.53%. wc:   1.01%. sc:   0.07%. cc:   0.33%. AOD: 0.1169. # of targets: 723/0.  1 product(s) written. dc:  70.53%. wc:   1.01%. sc:   0.07%. cc:   0.33%. AOD: 0.1169. # of targets: 723/0. Number of cols 7855 do not match for UPDATE/MERGE mode (file: 9011). error flushing L2 products.
Flushing images to disc failed! Error in geometric module.

**Reply by davidfrantz** [reply]
Please delete everything within the output folder (DIR_LEVEL2) and try again.

**Reply by krishnakumarkkp** [reply]
deleted everything in the 'level2' folder and tried. But again got the same result. Anyway I got something in the level2 output folder. Will check it in detail. But I found nothing in the log and temp folder.

update: Tried with a different image and got the same error. "# of targets: 723/0. Number of cols 7855 do not match for UPDATE/MERGE mode (file: 9011). error flushing L2 products. Flushing images to disc failed! Error in geometric module."

update: I removed the directory 'level2' and created new one using 'mkdir'. Now got the result: dc:  70.53%. wc:   1.01%. sc:   0.07%. cc:   0.33%. AOD: 0.1169. # of targets: 723/0.  1 product(s) written. Success! Processing time: 01 mins 14 secs.


Thanks to all for the wonderful support!
